[Feat] Caching 어노테이션을 이용하여 캐시처리

ParkJin0814 opened the pull request:
AOP를 이용하여 상품조회와 인기검색어 관련 책임분리
- 상품검색은 상품리스트만 반환하도록 책임분리
- AOP를 이용하여 후처리로 인기검색어 카운트수 증가처리

Cacheable, CacheEvict 어노테이션을 이용하여 캐시처리
- 기존 직접 저장하는 방식이 아닌 어노테이션으로 AOP 방식으로 처리
- CacheEvict를 이용하여 수정, 삭제같은 값이 변환되는 경우 캐시삭제 
- 상품의 경우 수정이없으면 캐시삭제가 필요없을거같아 TTL 별도 설정하지않음

인기검색어의 경우 검색시 캐시 삭제
- 인기검색어 캐시의 TTL을 5분으로 설정하여 검색이 없어도 5분후에 다시 집계

ParkJin0814 committed:
feat: Caching 어노테이션을 이용하여 캐시처리

AOP를 이용하여 상품조회와 인기검색어 관련 책임분리
Cacheable, CacheEvict 어노테이션을 이용하여 캐시처리

인기검색어의 경우 검색시 캐시 삭제
캐시의 TTL을 5분으로 설정하여 검색이 없어도 5분후에 다시 집계

ParkJin0814 committed:
feat: 오더리스트 캐시적용

roog23 commented on src/main/java/com/example/oneplusone/domain/search/service/SearchService.java:
```diff
@@ -24,121 +24,43 @@
      */
-    public void createPageCache(Page<ProductResponse> productResponses, String search, Pageable pageable, Long userId) {
-        setKey(search, pageable, userId);
+    @CacheEvict(value = CacheKeyConstants.POPULAR_SEARCH)
```
이 작업을 수행하는 이유가 뭔가요?

ParkJin0814 replied:
인기검색어 기준 검색 카운팅시 
인기검색어 목록 조회캐시에 변화가 있을수 있어서 캐시를 날렸습니다.

지금 생각해보니 변화를 체크하고 날리는것도 괜찮을거 같네여

roog23 replied:
private final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyyMMddHH");
String popular_search_time = "popular_search:" + LocalDateTime.now().format(formatter); 인기 검색어 조회 할 때 키 값을 현재의 시간을 기준으로 하여서 7시로 넘어가면 6시 캐시가 아니라 7시 캐시를 조회합니다. 
6시 캐시는 TTL에 의해 자동으로 날라가기 때문에 Evict작업이 필요없을 것이라고 생각됩니다.

roog23 commented on src/main/java/com/example/oneplusone/global/config/RedisConfig.java:
```diff
@@ -48,13 +43,20 @@ public RedisTemplate<String, Long> pageTotalElements(RedisConnectionFactory conn
+        return RedisCacheManager.builder(connectionFactory)
+                .cacheDefaults(config)
+                .withCacheConfiguration(
+                        // 인기검색어의 경우 TTL 5분 설정
```
인기 검색어 TTL 5분으로 설정하셨는데 위에서 인기 검색어는 계속 1시간 단위 기록인데 5분으로 설정하신 이유가 있나요?

ParkJin0814 replied:
구현되어있는 인기검색어의 기준 1시간단위로 되어있는데 
TTL 시간을 길게 가져갈 경우
예를 들어 6시 30분에 생성된 캐시가 7시가 넘어도 남아 있으면 잘못된 결과가 나올 수 있을 것 같다고 생각했습니다
스케줄러로 매 정각 초기화하는 방법도 있지만 TTL을 짧게 잡아 주기적으로 만료되게 하는 쪽이 로직추가 없이 간편해서 선택했습니다

roog23 replied:
위랑 같은 내용이긴 한데 저희가 실시간 검색을 1시간으로 진행하게 되면 TTL은 짧게 안잡아도 된다고 생각하긴 합니다.

roog23 commented on src/main/java/com/example/oneplusone/domain/orders/service/OrderService.java:
```diff
@@ -23,8 +26,16 @@ public class OrderService {
     @Transactional
+    @Caching(evict = {
+            // 검색캐시 전부제거
+            @CacheEvict(value = CacheKeyConstants.SEARCH_PRODUCT, allEntries = true),
```
이건 고려? 의논할 점.
변경되는 상품 외에도 모든 캐시를 삭제하는데 과한 삭제가 이루어지는 부분이 아닌가?

ryusu12 replied:
전체 조회 시, 여러 목록들에 걸쳐서 변경된 상품이 존재할 수 있다고 생각해서, 상품 전체 조회 캐시는 전부 지워주는 게 좋다고 생각합니다.
아니면, 변경된 상품이 포함된 목록만 삭제하는 로직이 별도로 구현해야 하는데 복잡하지 않을까요?

ParkJin0814 replied:
수연님이 말씀해주신거처럼 별도로 로직을 구현하면 복잡해질거같아서 크게 고려해보지 않았던거 같습니다

ParkJin0814 committed:
fix: 상품목록 조회 캐싱 오류수정

오류이유
 - GenericJackson2JsonRedisSerializer가 Page를 역질렬화하지 못해 캐시 복원 실패
 해결
 - 반환 타입을 Page에서 DTO로 변경하여 해결

ParkJin0814 committed:
fix: 사용하지않는 엔티티 제거

ParkJin0814 committed:
fix: 서치캐시관련 문제 해결

ParkJin0814 committed:
fix: 인기검색어목록 캐시 삭제

ParkJin0814 committed:
fix: 오더스 페이지 직렬화 문제해결

ryusu12 reviewed (approved):
충돌난거 해결만 해주시면 될것같습니다!

ParkJin0814 committed:
feat: 상품을 생성할때도 리스트에 변화가 있어 캐시삭제

ParkJin0814 committed:
Merge branch 'develop' into feat/caching-pj
